wgsl/index.bs

```diff
@@ -3762,36 +3752,262 @@ See [[#declaration-and-scope]].
+enclosing the declaration.
+A function-scope declaration is a [=dynamic context=].
+
+Variable and value declarations have a similar overall syntax:
+<xmp highlight=rust>
+  // Specific value declarations.
+               const    name [: type]  = initializer ;
+  [attribute]? override name [: type] [= initializer];
+               let      name [: type]  = initializer ;
+
+  // General variable form.
+  [attribute]* var[<address_space[, access_mode]>] name [: type] [= initializer];
+
+  // Specific variable declarations.
+  // Function scope.
+               var[<function>] name [: type] [= initializer];
+
+  // Module scope.
+               var<private>    name [: type] [= initializer];
+               var<workgroup>  name : type;
+  [attribute]+ var<uniform>    name : type;
+  [attribute]+ var             name : texture_type;
+  [attribute]+ var             name : sampler_type;
+  [attribute]+ var<storage[, access_mode]> name : type;
+</xmp>
 
-[SHORTNAME] authors can declare names for immutable values using a <dfn noexport>value declaration</dfn> which are either:
+Each such declaration [=shader-creation error|must=] have an explicitly
+specified type, an initializer, or both.
+Each such declaration determines the type for the associated data value, known
+as the <dfn noexport>effective-value-type</dfn> for the declaration.
+The effective-value-type of the declaration is:
+* The declared type, if explicitly specified.
+* Otherwise, if the initializer expression has type `T`:
+    * For a `const` declaration, the effective-value-type is `T` itself.
+    * For a `override`, `let`, or `var` declaration, the effective-value-type
+        is the [=concretization=] of `T`.
+
+Each kind of value or variable declaration may place additional constraints on
+the form of the initializer expression, if present, and on the
+effective-value-type.
 
-  * a [=let-declaration=], or
-  * an [=override-declaration=], or
-  * a [=const-declaration=].
+<table class='data'>
+<caption>
+  Variable and Value Declaration Feature Summary.
+</caption>
+<thead>
+  <tr><th>Declaration
+      <th>Mutability
+      <th>Scope
+      <th>Type<sup>1</sup>
+      <th>Initializer Support
+      <th>Initializer Expression<sup>2</sup>
+      <th>Part of Resource Interface
+</thead>
+<tr><td>[=const-declaration|const=]
+    <td>Immutable
+    <td>[=module scope|Module=] or [=function scope|function=]
+    <td>[=type/concrete|Concrete=] or [=type/abstract|abstract=] [=constructible=]
+    <td>Required
+    <td>[=const-expression=]
+    <td>No
+
+<tr><td>[=override-declaration|override=]
+    <td>Immutable
+    <td>[=module scope|Module=]
+    <td>[=type/concrete|Concrete=] [=scalar=]
+    <td>Optional<sup>3</sup>
+    <td>[=override-expression=]
+    <td>No<sup>4</sup>
+
+<tr><td>[=let-declaration|let=]
+    <td>Immutable
+    <td>[=function scope|Function=]
+    <td>[=type/concrete|Concrete=] [=constructible=] or [=pointer type=]
+    <td>Required
+    <td>[=runtime expression=]
```

Comment:
By the definition below, this excludes const-expression and override-expression. That's not right.
This would be runtime-expression, const-expression or override-expression.

We could go three ways: 
1. expand runtime to include the others and tweak the language below. I tried that and it was a lot of boilerplate.
2. Add yet another concept 'unrestricted' expression which is all three kinds: runtime, override, and const expression. But that's too suggestive that *other* constraints on the type don't apply. So it's a possible confusion, and doesn't have much payoff.
3. add more cases here for 'let' and 'var'.   This is simpler.

I like 3.



Suggest, both here and at 'var<function>':

> const-expression, pipeline-expression, or runtime-expression

Technically this really only needs pipeline-expression or runtime-expression, but I'm sure it's easier to grok at a glance if we list all three.